### PR TITLE
[SofaImGui] Fix window positioning and state persistence issues

### PR DESCRIPTION
**Pull request by lamriaimen**
- Implemented mechanism to persist window positions and states across application runs.
- Inject `first_run.txt` in the `builder/bin` folder once first run is completed.
- Added file check for `first_run.txt` to  set ImGui docking data just in the first run.
- Fixed issue with window positions not being saved, ensuring they remain in place across sessions.
- Resolved lag issue with windows not being properly pinned due to size differences with the interface.

### CODE REVIEW

File: SofaImGui/src/SofaImGui/ImGuiGUIEngine.cpp [hunk 1]
```diff
@@ -256,21 +256,32 @@ void ImGuiGUIEngine::startFrame(sofaglfw::SofaGLFWBaseGUI* baseGUI)
...
         ImGui::DockBuilderDockWindow(windowNameViewport, dockspace_id);
         ImGui::DockBuilderFinish(dockspace_id);
+
+        static bool isViewportWindowOpen = true;
+        winManager.createWindowStateFile(winManager.fileNameViewPort);
+
+        static bool isSceneGraphWindowOpen = true;
+        winManager.createWindowStateFile(winManager.fileNameSceneGraph);
+
+        static bool isLogWindowOpen = true;
```
Comment (alxbilger) [inline SofaImGui/src/SofaImGui/ImGuiGUIEngine.cpp:266]: What is the purpose of the 3 static boolean variables in this scope?

File: SofaImGui/src/SofaImGui/ImGuiGUIEngine.cpp [hunk 2]
```diff
@@ -614,7 +652,7 @@ void ImGuiGUIEngine::beforeDraw(GLFWwindow*)
     if (!m_fbo)
     {
         m_fbo = std::make_unique<sofa::gl::FrameBufferObject>();
-        m_currentFBOSize = {500, 500};
+        m_currentFBOSize = {50, 50};
```
Comment (alxbilger) [inline SofaImGui/src/SofaImGui/ImGuiGUIEngine.cpp:655]: Can you explain why you changed the values here?

File: SofaImGui/src/SofaImGui/ImGuiGUIEngine.h [hunk 3]
```diff
@@ -29,12 +29,14 @@
 #include <imgui.h>
 #include <sofa/simulation/Node.h>
 #include <SimpleIni.h>
+#include "windows/WindowsManager.h"
```
Comment (alxbilger) [inline SofaImGui/src/SofaImGui/ImGuiGUIEngine.h:32]: I am not sure you need to include this file in the header. Can you include it only in the cpp file?

File: SofaImGui/src/SofaImGui/windows/Components.cpp [hunk 4]
```diff
@@ -39,17 +37,19 @@
 #include <sofa/gui/common/BaseGUI.h>
 #include <sofa/simulation/graph/DAGNode.h>
 
-
+#include "WindowsManager.h"
 #include "Components.h"
 
 namespace windows
 {
+    extern WindowsManager winManager;
```
Comment (alxbilger) [inline SofaImGui/src/SofaImGui/windows/Components.cpp:45]:
```suggestion
    WindowsManager winManager;
```

File: SofaImGui/src/SofaImGui/windows/WindowsManager.cpp [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        if (!checkIfWindowFileExist(windowType)) {
+            std::ofstream outfile(windowType + ".txt");
+            outfile << "This file marks that the " << windowType << " window will open in the next run";
+            outfile.close();
+        }
+    }
+
+
+    void WindowsManager::removeWindowStateFile(const std::string &windowType) {
+        if (checkIfWindowFileExist(windowType)) {
```
Comment (alxbilger) [inline SofaImGui/src/SofaImGui/windows/WindowsManager.cpp:57]: [Allman style](https://en.wikipedia.org/wiki/Indentation_style#Allman_style) please. Curly brace on the next line

**Review by alxbilger — CHANGES_REQUESTED** [review]
You designed `WindowsManager` to be a class, but you don't use instances of this class properly. For example, you don't use the same instance in ImGuiGUIEngine.cpp and in Components.cpp.
But the good news is that `WindowsManager` does not need to be a class. Indeed, it does not store anything internally. I suggest to create only free functions in the namespace `windows`.

I also would like to know if the .ini file created by imgui does not store also the state of a window (open or close). Just wanted to know if you really need to store it by yourself using `checkIfWindowFileExist`.

File: SofaImGui/src/SofaImGui/windows/Profiler.cpp [hunk 6]
```diff
@@ -23,119 +23,96 @@
 #include <imgui_internal.h> //imgui_internal.h is included in order to use the DockspaceBuilder API (which is still in development)
```
Comment (alxbilger) [inline SofaImGui/src/SofaImGui/windows/Profiler.cpp:23]: you changed the format of the file. It's difficult to review as the true differences are mixed with format changes. Could you revert the format changes only?

**Review by alxbilger — CHANGES_REQUESTED** [review]

File: SofaImGui/src/SofaImGui/ImGuiGUIEngine.cpp [hunk 7]
```diff
@@ -81,13 +80,26 @@
...
+    WindowState winManagerProfiler("profiler.txt");
+    WindowState winManagerSceneGraph("scenegraph.txt");
+    WindowState winManagerPerformances("performances.txt");
+    WindowState winManagerDisplayFlags("displayflags.txt");
+    WindowState winManagerPlugins("plugins.txt");
+    WindowState winManagerComponents("components.txt");
+    WindowState winManagerLog("log.txt");
+    WindowState winManagerSettings("settings.txt");
+    WindowState winManagerViewPort("viewport.txt");
+    WindowState firstRunState("firstrun.txt");
```
Comment (alxbilger) [inline SofaImGui/src/SofaImGui/ImGuiGUIEngine.cpp:99]: I recommend to define them as class members of `ImGuiGUIEngine`

**Review by alxbilger — CHANGES_REQUESTED** [review]

**Comment by lamriaimen** [comment]
This Pr needs [https://github.com/sofa-framework/SofaGLFW/pull/113](url) to be merged before It merges

**Comment by lamriaimen** [comment]
@alxbilger  https://github.com/sofa-framework/SofaGLFW/pull/114 didn't fix the problem of the non  destructor call

**Comment by lamriaimen** [comment]
@alxbilger it's working. the Pr https://github.com/sofa-framework/SofaGLFW/pull/117 fixes the problem and it needs to be merged before this one as you recommended.

**Review by alxbilger — APPROVED** [review]

**Pull request merged**